### PR TITLE
Alter TailRec to have map and flatMap

### PR DESCRIPTION
**Pull request by folone**
I did not get any replies on the [mailing list](https://groups.google.com/forum/#!topic/scala-internals/2Thm0YE_7zY). A PR probably has better chances to be noticed.
Here is an implementation of `map` and `flatMap` for `TailRec`, using technique, described by @runarorama (http://apocalisp.wordpress.com/2011/10/26/tail-call-elimination-in-scala-monads/).

Review by @phaller and @axel22 please (as per [policy](http://docs.scala-lang.org/scala/pull-request-policy.html)).
Also ping @odersky, as it is his code being updated.

### CODE REVIEW

File: src/library/scala/util/control/TailCalls.scala [hunk 1]
```diff
@@ -31,14 +45,31 @@ object TailCalls {
   /** This class represents a tailcalling computation
    */
   abstract class TailRec[+A] {
+    def map[B](f: A => B): TailRec[B] =
+      flatMap(a => Call(() => Done(f(a))))
+    def flatMap[B](f: A => TailRec[B]): TailRec[B] =
+      Cont(this, f)
```
Comment (runarorama) [inline src/library/scala/util/control/TailCalls.scala]: You should avoid creating left-leaning towers of `Cont` here. Running the trampoline with a tower like `Cont(Cont(Cont(Cont(...` will require your `stack` to be unnecessarily large.

Comment (runarorama) [inline src/library/scala/util/control/TailCalls.scala]: Pattern match on `this` and reassociate `Cont` to the right.

**Comment by retronym** [comment]
LGTM. @JamesIry perhaps you can also take a look?

**Comment by gkossakowski** [comment]
LGTM.

I stole review because I'm guessing that both @phaller and @axel22 are busy.

@folone: I'm sorry it took so long to review. We've been busy keeping up with PR traffic which is probably a good problem to have! Thanks a lot for your contribution.

**Comment by folone** [comment]
@gkossakowski Yeah, I've seen that a lot of work has been done to wrap up M5. Congrats with the release! :+1:

**Pull request merged**